aries_vcx/src/protocols/connection/generic/mod.rs

```diff
@@ -0,0 +1,481 @@
+mod conversions;
+mod thin_state;
+
+use std::sync::Arc;
+
+use messages::{a2a::A2AMessage, diddoc::aries::diddoc::AriesDidDoc, protocols::connection::invite::Invitation};
+
+pub use self::thin_state::{State, ThinState};
+
+use crate::{
+    errors::error::{AriesVcxError, AriesVcxErrorKind, VcxResult},
+    plugins::wallet::base_wallet::BaseWallet,
+    protocols::connection::{
+        invitee::states::{
+            complete::Complete as InviteeComplete, initial::Initial as InviteeInitial,
+            invited::Invited as InviteeInvited, requested::Requested as InviteeRequested,
+            responded::Responded as InviteeResponded,
+        },
+        inviter::states::{
+            complete::Complete as InviterComplete, initial::Initial as InviterInitial,
+            invited::Invited as InviterInvited, requested::Requested as InviterRequested,
+            responded::Responded as InviterResponded,
+        },
+        pairwise_info::PairwiseInfo,
+        trait_bounds::{TheirDidDoc, ThreadId},
+    },
+    transport::Transport,
+};
+
+use super::{basic_send_message, trait_bounds::BootstrapDidDoc};
+
+/// A type that can encapsulate a [`Connection`] of any state.
+/// While mainly used for deserialization, it exposes some methods for retrieving
+/// connection information.
+///
+/// However, using methods directly from [`Connection`], if possible, comes with certain
+/// benefits such as being able to obtain an [`AriesDidDoc`] directly (if the state contains it)
+/// and not an [`Option<AriesDidDoc>`] (which is what [`GenericConnection`] provides).
+///
+/// [`GenericConnection`] implements [`From`] for all [`Connection`] states and
+/// [`Connection`] implements [`TryFrom`] from [`GenericConnection`], with the conversion failing
+/// if the [`GenericConnection`] is in a different state than the requested one.
+/// This is also the mechanism used for direct deserialization of a [`Connection`].
+///
+/// Because a [`TryFrom`] conversion is fallible and consumes the [`GenericConnection`], a thin [`State`]
+/// can be retrieved through [`GenericConnection::state`] method at runtime. In that case, a more dynamic conversion
+/// could be done this way:
+///
+/// ``` ignore
+/// // Assume the `con` variable stores a `GenericConnection`:
+///
+/// let initial_connections = Vec::new();
+/// let completed_connections = Vec::new();
+///
+/// // Unwrapping after the match is sound
+/// // because we can guarantee the conversion will work
+/// match con.state() {
+///     State::Invitee(Stage::Initial) => initial_connections.push(con.try_into().unwrap()),
```

Comment:
Just a general comment to make sure you circle back to this docco - i.e. it should be `ThinState::Invitee(State::Initial)` I believe. other parts of this docco make need to be double checked as well

Reply:
Good catch. It took a bit to settle on some naming 😅 .

Fixed.